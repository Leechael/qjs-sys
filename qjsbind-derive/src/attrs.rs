use std::borrow::Cow;

#[derive(Copy, Clone)]
#[allow(clippy::enum_variant_names)]
pub enum RenameAll {
    LowerCase,
    UpperCase,
    PascalCase,
    CamelCase,
    SnakeCase,
    ScreamingSnakeCase,
    KebabCase,
    ScreamingKebabCase,
}

impl RenameAll {
    fn parse(lit: &syn::Lit) -> Result<RenameAll, syn::Error> {
        if let syn::Lit::Str(s) = &lit {
            match s.value().as_str() {
                "lowercase" => Ok(RenameAll::LowerCase),
                "UPPERCASE" => Ok(RenameAll::UpperCase),
                "PascalCase" => Ok(RenameAll::PascalCase),
                "camelCase" => Ok(RenameAll::CamelCase),
                "snake_case" => Ok(RenameAll::SnakeCase),
                "SCREAMING_SNAKE_CASE" => Ok(RenameAll::ScreamingSnakeCase),
                "kebab-case" => Ok(RenameAll::KebabCase),
                "SCREAMING-KEBAB-CASE" => Ok(RenameAll::ScreamingKebabCase),
                _ => Err(syn::Error::new_spanned(lit, "")),
            }
        } else {
            Err(syn::Error::new_spanned(lit, "rename expects a string"))
        }
    }
}

#[derive(Clone)]
pub enum TypeDefault {
    Implicit,
    Explicit(syn::ExprPath),
}

pub struct ContainerAttrs<'a> {
    ident: &'a syn::Ident,
    rename_all: Option<RenameAll>,
}

pub fn get_meta_items(attr: &syn::Attribute) -> syn::Result<Vec<syn::NestedMeta>> {
    if !attr.path.is_ident("qjsbind") {
        return Ok(Vec::new());
    }

    match attr.parse_meta() {
        Ok(syn::Meta::List(meta)) => Ok(meta.nested.into_iter().collect()),
        Ok(_) => Err(syn::Error::new_spanned(attr, "expected #[qjsbind(...)]")),
        Err(err) => Err(err),
    }
}

fn respan(stream: proc_macro2::TokenStream, span: proc_macro2::Span) -> proc_macro2::TokenStream {
    stream
        .into_iter()
        .map(|mut token| {
            if let proc_macro2::TokenTree::Group(g) = &mut token {
                *g = proc_macro2::Group::new(g.delimiter(), respan(g.stream(), span));
            }
            token.set_span(span);
            token
        })
        .collect()
}

fn get_lit_str(attr_name: &str, lit: &syn::Lit) -> syn::Result<String> {
    if let syn::Lit::Str(lit) = lit {
        Ok(lit.value())
    } else {
        Err(syn::Error::new_spanned(
            lit,
            format!(
                "expected attribute to be a string: `{} = \"...\"`",
                attr_name,
            ),
        ))
    }
}

fn parse_lit_into_expr_path(attr_name: &str, lit: &syn::Lit) -> syn::Result<syn::ExprPath> {
    let string = get_lit_str(attr_name, lit)?;
    let token_stream = syn::parse_str(&string)?;
    syn::parse2(respan(token_stream, lit.span()))
}

impl<'a> ContainerAttrs<'a> {
    pub fn of(input: &'a syn::DeriveInput) -> syn::Result<ContainerAttrs<'a>> {
        let mut rv = ContainerAttrs {
            ident: &input.ident,
            rename_all: None,
        };

        for meta_item in input.attrs.iter().flat_map(get_meta_items).flatten() {
            if let syn::NestedMeta::Meta(meta) = meta_item {
                match &meta {
                    syn::Meta::NameValue(nv) if nv.path.is_ident("rename_all") => {
                        if rv.rename_all.is_some() {
                            return Err(syn::Error::new_spanned(
                                meta,
                                "duplicate rename_all attribute",
                            ));
                        }
                        rv.rename_all = Some(RenameAll::parse(&nv.lit)?);
                    }
                    _ => return Err(syn::Error::new_spanned(meta, "unsupported attribute")),
                }
            } else {
                return Err(syn::Error::new_spanned(meta_item, "unsupported attribute"));
            }
        }

        Ok(rv)
    }

    pub fn get_field_name(&self, field: &syn::Field) -> String {
        let name = field.ident.as_ref().unwrap().to_string();
        if let Some(rename_all) = self.rename_all {
            match rename_all {
                RenameAll::LowerCase | RenameAll::SnakeCase => name,
                RenameAll::UpperCase | RenameAll::ScreamingSnakeCase => name.to_ascii_uppercase(),
                RenameAll::PascalCase => {
                    let mut pascal = String::new();
                    let mut capitalize = true;
                    for ch in name.chars() {
                        if ch == '_' {
                            capitalize = true;
                        } else if capitalize {
                            pascal.push(ch.to_ascii_uppercase());
                            capitalize = false;
                        } else {
                            pascal.push(ch);
                        }
                    }
                    pascal
                }
                RenameAll::CamelCase => {
                    let mut camel = String::new();
                    let mut capitalize = false;
                    for ch in name.chars() {
                        if ch == '_' {
                            capitalize = true;
                        } else if capitalize {
                            camel.push(ch.to_ascii_uppercase());
                            capitalize = false;
                        } else {
                            camel.push(ch);
                        }
                    }
                    camel
                }
                RenameAll::KebabCase => name.replace("_", "-"),
                RenameAll::ScreamingKebabCase => name.replace("_", "-").to_ascii_uppercase(),
            }
        } else {
            name
        }
    }

    pub fn ident(&self) -> &syn::Ident {
        self.ident
    }
}

pub struct FieldAttrs<'a> {
    field: &'a syn::Field,
    rename: Option<String>,
    default: Option<TypeDefault>,
    as_bytes: bool,
}

impl<'a> FieldAttrs<'a> {
    pub fn of(field: &'a syn::Field) -> syn::Result<FieldAttrs<'a>> {
        let mut rv = FieldAttrs {
            field,
            rename: None,
            default: None,
            as_bytes: false,
        };

        for meta_item in field.attrs.iter().flat_map(get_meta_items).flatten() {
            if let syn::NestedMeta::Meta(meta) = meta_item {
                match &meta {
                    syn::Meta::NameValue(nv) if nv.path.is_ident("rename") => {
                        if rv.rename.is_some() {
                            return Err(syn::Error::new_spanned(
                                meta,
                                "duplicate rename attribute",
                            ));
                        }
                        rv.rename = Some(get_lit_str("rename", &nv.lit)?);
                    }
                    syn::Meta::NameValue(nv) if nv.path.is_ident("default") => {
                        if rv.default.is_some() {
                            return Err(syn::Error::new_spanned(
                                meta,
                                "duplicate default attribute",
                            ));
                        }
                        rv.default = Some(TypeDefault::Explicit(parse_lit_into_expr_path(
                            "default", &nv.lit,
                        )?));
                    }
                    syn::Meta::Path(path) if path.is_ident("as_bytes") => {
                        if rv.as_bytes {
                            return Err(syn::Error::new_spanned(
                                meta,
                                "duplicate as_bytes attribute",
                            ));
                        }
                        rv.as_bytes = true;
                    }
                    syn::Meta::Path(path) if path.is_ident("default") => {
                        if rv.default.is_some() {
                            return Err(syn::Error::new_spanned(
                                meta,
                                "duplicate default attribute",
                            ));
                        }
                        rv.default = Some(TypeDefault::Implicit);
                    }
                    _ => return Err(syn::Error::new_spanned(meta, "unsupported attribute")),
                }
            } else {
                return Err(syn::Error::new_spanned(meta_item, "unsupported attribute"));
            }
        }

        Ok(rv)
    }

    pub fn field(&self) -> &syn::Field {
        self.field
    }

    pub fn name(&self, container_attrs: &ContainerAttrs) -> Cow<'_, str> {
        self.rename
            .as_deref()
            .map(Cow::Borrowed)
            .unwrap_or_else(|| container_attrs.get_field_name(self.field).into())
    }

    pub fn as_bytes(&self) -> bool {
        self.as_bytes
    }

    pub fn decoder_fn(&self, crate_qjsbind: &syn::Ident) -> syn::Path {
        if self.as_bytes {
            syn::parse_quote!(#crate_qjsbind::decode_as_bytes)
        } else {
            syn::parse_quote!(FromJsValue::from_js_value)
        }
    }

    pub fn default_fn(&self) -> Option<syn::ExprPath> {
        match &self.default {
            Some(TypeDefault::Implicit) => Some(syn::parse_quote!(Default::default)),
            Some(TypeDefault::Explicit(path)) => Some(path.clone()),
            None => None,
        }
    }
}
