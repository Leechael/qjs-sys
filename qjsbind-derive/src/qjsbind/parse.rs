use syn::{FnArg, Pat, Type};

use super::*;

macro_rules! extract_qjs_attrs {
    ($item: ident) => {{
        let (qjs_attrs, attrs) = partition_attributes($item.attrs.clone())?;
        $item.attrs = attrs;
        if (qjs_attrs.is_empty()) {
            None
        } else {
            Some(qjs_attrs)
        }
    }};
}

impl Class {
    fn from_struct(item_struct: &mut ItemStruct, js_crate: &syn::Path) -> Result<Option<Self>> {
        let Some(qjs_attrs) = extract_qjs_attrs!(item_struct) else {
            return Ok(None);
        };
        item_struct
            .attrs
            .push(syn::parse_quote!(#[derive(#js_crate::GcMark)]));
        let name = item_struct.ident.clone();
        let attrs = ClassAttrs::from_attributes(&qjs_attrs)?;

        let mut fields = vec![];
        for field in &mut item_struct.fields.iter_mut() {
            let qjs_property = DerivedProperty::from_field(field)?;
            fields.push(ClassField {
                field: field.clone(),
                qjs_property,
            });
        }

        Ok(Some(Self {
            name,
            methods: Vec::new(),
            attrs,
            fields,
            constructor: None,
        }))
    }

    fn validate(&self) -> Result<()> {
        Ok(())
    }
}

impl ClassAttrs {
    fn from_attributes(attrs: &[Attribute]) -> Result<Self> {
        let mut js_name = None;
        let mut rename_all = None;
        let mut is_class = false;

        for attr in attrs {
            if attr.path().is_ident("qjs") {
                attr.parse_nested_meta(|meta| {
                    if meta.path.is_ident("class") {
                        is_class = true;
                        if meta.input.is_empty() {
                            return Ok(());
                        }
                        meta.parse_nested_meta(|meta| {
                            if meta.path.is_ident("js_name") {
                                ensure_none!(js_name, meta.path, "duplicate `js_name` attribute");
                                js_name = Some(meta.value()?.parse::<LitStr>()?);
                            } else if meta.path.is_ident("rename_all") {
                                let lit_rename_all = meta.value()?.parse::<LitStr>()?;
                                ensure_none!(
                                    rename_all,
                                    meta.path,
                                    "duplicate `rename_all` attribute"
                                );
                                rename_all = Some(RenameAll::parse(&lit_rename_all)?);
                            } else {
                                syn_bail!(meta.path, "unknown attribute");
                            }
                            Ok(())
                        })?;
                    } else {
                        syn_bail!(meta.path, "unknown attribute");
                    }
                    Ok(())
                })?;
            }
        }

        if !is_class {
            syn_bail!(attrs[0], "expected `class` attribute");
        }
        Ok(Self {
            js_name,
            rename_all: rename_all.or(Some(RenameAll::CamelCase)),
        })
    }
}

impl DerivedProperty {
    fn from_field(field: &mut Field) -> Result<Option<Self>> {
        let Some(qjs_attrs) = extract_qjs_attrs!(field) else {
            return Ok(None);
        };
        let attrs = FieldAttrs::from_attributes(&qjs_attrs)?;
        let Some(name) = field.ident.clone() else {
            syn_bail!(field, "expected named field");
        };
        let ty = field.ty.clone();
        Ok(Some(Self { name, ty, attrs }))
    }
}

impl FieldAttrs {
    fn from_attributes(attrs: &[Attribute]) -> Result<Self> {
        let mut js_name = None;
        let mut getter = None;
        let mut setter = None;

        for attr in attrs {
            if attr.path().is_ident("qjs") {
                attr.parse_nested_meta(|meta| {
                    let Some(ident) = meta.path.get_ident() else {
                        syn_bail!(meta.path, "expected an identifier");
                    };
                    match ident.to_string().as_str() {
                        "getter" => {
                            getter = Some(ident.clone());
                        }
                        "setter" => {
                            setter = Some(ident.clone());
                        }
                        "js_name" => {
                            ensure_none!(js_name, meta.path, "duplicate `js_name` attribute");
                            js_name = Some(meta.value()?.parse::<LitStr>()?);
                        }
                        _ => {
                            syn_bail!(meta.path, "unknown attribute");
                        }
                    }
                    Ok(())
                })?;
            }
        }

        Ok(Self {
            js_name,
            getter,
            setter,
        })
    }
}

enum FnItem {
    Constructor(Constructor),
    Method(Method),
}

fn parse_fn_item(item_fn: &mut ImplItemFn) -> Result<Option<FnItem>> {
    let Some(qjs_attrs) = extract_qjs_attrs!(item_fn) else {
        return Ok(None);
    };
    match parse_fn_attributes(&qjs_attrs)? {
        FnAttrs::Method(attrs) => Method::from_item_fn(item_fn, attrs)
            .map(FnItem::Method)
            .map(Some),
        // Currently, None means it's a constructor
        FnAttrs::Constructor(attrs) => Constructor::from_item_fn(item_fn, attrs)
            .map(FnItem::Constructor)
            .map(Some),
    }
}

impl Method {
    fn from_item_fn(item_fn: &mut ImplItemFn, attrs: MethodAttrs) -> Result<Self> {
        let name = item_fn.sig.ident.clone();
        let args = parse_fn_args(item_fn.sig.inputs.iter_mut())?;
        if let Some(receiver) = &args.receiver {
            if !receiver.is_ref {
                syn_bail!(receiver.token, "expected a reference receiver");
            }
        }
        let is_static = args.receiver.is_none();
        let return_ty = item_fn.sig.output.clone();
        // validate
        match &attrs.fn_type {
            MethodType::Getter => {
                if !args.args.is_empty() {
                    syn_bail!(attrs.marker_token, "getter method cannot take arguments");
                }
            }
            MethodType::Setter => {
                if args.args.len() != 1 {
                    syn_bail!(
                        attrs.marker_token,
                        "setter method must take exactly one argument"
                    );
                }
            }
            _ => (),
        }
        Ok(Self {
            name,
            attrs,
            is_mut: args.receiver.as_ref().map_or(false, |r| r.is_mut),
            is_static,
            args,
            return_ty,
        })
    }
}

enum FnAttrs {
    Constructor(ConstructorAttrs),
    Method(MethodAttrs),
}

fn parse_fn_attributes(attrs: &[Attribute]) -> Result<FnAttrs> {
    let mut js_name = None;
    let mut fn_type = None;

    for attr in attrs {
        if attr.path().is_ident("qjs") {
            attr.parse_nested_meta(|meta| {
                let Some(ident) = meta.path.get_ident() else {
                    syn_bail!(meta.path, "expected an identifier");
                };
                match ident.to_string().as_str() {
                    "method" => {
                        fn_type = Some((MethodType::Method, ident.clone()));
                    }
                    "getter" => {
                        fn_type = Some((MethodType::Getter, ident.clone()));
                    }
                    "setter" => {
                        fn_type = Some((MethodType::Setter, ident.clone()));
                    }
                    "constructor" => {
                        fn_type = Some((MethodType::Constructor, ident.clone()));
                    }
                    "js_name" => {
                        ensure_none!(js_name, meta.path, "duplicate `js_name` attribute");
                        js_name = Some(meta.value()?.parse::<LitStr>()?);
                    }
                    _ => {
                        syn_bail!(meta.path, "unknown attribute");
                    }
                }
                Ok(())
            })?;
        }
    }

    let Some((fn_type, marker_token)) = fn_type else {
        syn_bail!(
            attrs[0],
            "expected exactly one of `getter`, `setter`, `method`, or `constructor"
        );
    };
    match fn_type {
        MethodType::Constructor => {
            if js_name.is_some() {
                syn_bail!(js_name, "constructor cannot have `js_name` attribute");
            }
            Ok(FnAttrs::Constructor(ConstructorAttrs { marker_token }))
        }
        _ => Ok(FnAttrs::Method(MethodAttrs {
            js_name,
            fn_type,
            marker_token,
        })),
    }
}

fn parse_fn_args<'a>(mut inputs: impl Iterator<Item = &'a mut FnArg>) -> Result<Args> {
    let mut args = Args::default();
    let mut next = inputs.next();
    if let Some(FnArg::Receiver(receiver)) = next {
        args.receiver = Some(ArgSelf {
            is_ref: receiver.reference.is_some(),
            is_mut: receiver.mutability.is_some(),
            token: receiver.clone(),
        });
        next = inputs.next();
    }
    while let Some(arg) = next {
        let FnArg::Typed(pat) = arg else {
            syn_bail!(arg, "expected a typed argument");
        };
        let ident = match &*pat.pat {
            Pat::Ident(ident) => ident.ident.clone(),
            _ => {
                syn_bail!(pat.pat, "expected an identifier");
            }
        };
        let mut from_context = None;
        if let Some(qjs_attrs) = extract_qjs_attrs!(pat) {
            for attr in qjs_attrs {
                attr.parse_nested_meta(|meta| {
                    if meta.path.is_ident("from_context") {
                        from_context = meta.path.get_ident().cloned();
                    } else {
                        syn_bail!(meta.path, "unknown attribute");
                    }
                    Ok(())
                })?
            }
        }
        args.args.push(Arg {
            name: ident,
            ty: *pat.ty.clone(),
            from_context,
        });
        next = inputs.next();
    }
    Ok(args)
}

impl Constructor {
    fn from_item_fn(item_fn: &mut ImplItemFn, attrs: ConstructorAttrs) -> Result<Self> {
        let name = item_fn.sig.ident.clone();
        let args = parse_fn_args(item_fn.sig.inputs.iter_mut())?;
        if args.receiver.is_some() {
            syn_bail!(item_fn.sig.inputs, "constructor cannot take `self`");
        }
        Ok(Self { name, args, attrs })
    }
}

impl Mod {
    pub(crate) fn from_mod(item_mod: &mut ItemMod, js_crate: Option<Path>) -> Result<Self> {
        let js_crate = match js_crate {
            Some(js_crate) => js_crate,
            None => crate::find_crate_name("qjsbind")?.into(),
        };
        let mut classes = BTreeMap::new();
        if let Some((_, ref mut items)) = item_mod.content {
            for item in items {
                match item {
                    Item::Struct(item_struct) => {
                        let Some(class) = Class::from_struct(item_struct, &js_crate)? else {
                            continue;
                        };
                        classes.insert(class.name.to_string(), class);
                    }
                    Item::Impl(item_impl) => {
                        if item_impl.trait_.is_some() {
                            continue;
                        }
                        let Type::Path(ty) = item_impl.self_ty.as_ref() else {
                            continue;
                        };
                        let Some(ident) = ty.path.get_ident() else {
                            continue;
                        };
                        let Some(for_class) = classes.get_mut(&ident.to_string()) else {
                            continue;
                        };

                        for item in &mut item_impl.items {
                            match item {
                                syn::ImplItem::Fn(item_method) => {
                                    let Some(method) = parse_fn_item(item_method)? else {
                                        continue;
                                    };
                                    match method {
                                        FnItem::Constructor(constructor) => {
                                            ensure_none!(
                                                for_class.constructor,
                                                constructor.attrs.marker_token,
                                                "Only one constructor is allowed per class"
                                            );
                                            for_class.constructor = Some(constructor);
                                        }
                                        FnItem::Method(method) => {
                                            for_class.methods.push(method);
                                        }
                                    }
                                }
                                _ => {}
                            }
                        }
                    }
                    _ => {}
                }
            }
        }
        for cls in classes.values() {
            cls.validate()?;
        }
        Ok(Self { classes, js_crate })
    }
}

pub fn partition_attributes<I>(attrs: I) -> Result<(Vec<Attribute>, Vec<Attribute>)>
where
    I: IntoIterator<Item = Attribute>,
{
    use itertools::{Either, Itertools as _};
    let (qjs_attrs, others) = attrs.into_iter().partition_map(|attr| {
        if attr.path().is_ident("qjs") {
            Either::Left(attr)
        } else {
            Either::Right(attr)
        }
    });
    Ok((qjs_attrs, others))
}
