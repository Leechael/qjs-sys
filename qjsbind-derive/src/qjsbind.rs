//! This module contains the `qjsbind` attribute macro implementation.
//!

use proc_macro2::{Span, TokenStream};
use syn::spanned::Spanned;
use std::collections::BTreeMap;
use syn::parse::Parser;
use syn::{
    Attribute, Field, Ident, ImplItemFn, Item, ItemMod, ItemStruct, LitStr, Path, Receiver, Result,
    Type,
};
use template_quote::{quote, ToTokens};

use crate::attrs::RenameAll;

mod codegen;
mod parse;

struct Mod {
    js_crate: Path,
    classes: BTreeMap<String, Class>,
}

struct Class {
    name: Ident,
    constructor: Option<Constructor>,
    methods: Vec<Method>,
    fields: Vec<ClassField>,
    attrs: ClassAttrs,
}

struct ClassField {
    field: Field,
    qjs_property: Option<DerivedProperty>,
}

impl ClassField {
    fn span(&self) -> Span {
        self.field.span()
    }
}

struct ClassAttrs {
    js_name: Option<LitStr>,
    rename_all: Option<RenameAll>,
}

struct DerivedProperty {
    name: Ident,
    ty: Type,
    attrs: FieldAttrs,
}

struct FieldAttrs {
    js_name: Option<LitStr>,
    getter: Option<Ident>,
    setter: Option<Ident>,
}

struct ArgSelf {
    is_ref: bool,
    is_mut: bool,
    token: Receiver,
}

#[derive(Default)]
struct Args {
    receiver: Option<ArgSelf>,
    args: Vec<Arg>,
}

struct Arg {
    name: Ident,
    ty: Type,
    from_context: Option<Ident>,
}

struct Constructor {
    name: Ident,
    args: Args,
    attrs: ConstructorAttrs,
}

struct ConstructorAttrs {
    marker_token: Ident,
}

struct Method {
    name: Ident,
    args: Args,
    return_ty: syn::ReturnType,
    is_mut: bool,
    is_static: bool,
    attrs: MethodAttrs,
}

struct MethodAttrs {
    js_name: Option<LitStr>,
    fn_type: MethodType,
    marker_token: Ident,
}

#[derive(Debug, Clone, PartialEq, Eq)]
enum MethodType {
    Getter,
    Setter,
    Method,
    Constructor,
}

pub(crate) fn patch(config: TokenStream, input: TokenStream) -> TokenStream {
    match patch_or_err(config, input) {
        Ok(tokens) => tokens,
        Err(err) => err.to_compile_error(),
    }
}

fn patch_or_err(config: TokenStream, input: TokenStream) -> Result<TokenStream> {
    let mut js_crate = None;
    syn::meta::parser(|meta| {
        if meta.path.is_ident("js_crate") {
            js_crate = Some(meta.value()?.parse::<Path>()?);
        }
        Ok(())
    })
    .parse2(config)?;
    let mut the_mod: ItemMod = syn::parse2(input)?;
    let qjs_mod = Mod::from_mod(&mut the_mod, js_crate)?;
    let ItemMod {
        attrs,
        vis,
        unsafety,
        mod_token,
        ident,
        content,
        semi,
    } = the_mod;

    let Some((_brace, content)) = content else {
        syn_bail!(semi, "expected a module with content");
    };

    Ok(quote! {
        #(#attrs)*
        #vis #unsafety #mod_token #ident {
            #(#content)*
            #qjs_mod
        }
    })
}

#[test]
fn show_tokens() {
    let tokens = quote! {
        mod native_classes {
            use js::IntoNativeObject as _;

            use super::{KeyGenAlgorithm, Result};

            #[qjs(class(rename_all = "camelCase"))]
            pub struct CryptoKey {}

            impl CryptoKey {
                #[qjs(constructor)]
                pub fn new(inner: CryptoKey) -> Result<Self> {
                    Ok(inner)
                }
            }
        }
    };
    let patched = patch(quote!(js_crate = js), tokens);
    insta::assert_snapshot!(rustfmt_snippet::rustfmt(&patched.to_string()).unwrap());
}
