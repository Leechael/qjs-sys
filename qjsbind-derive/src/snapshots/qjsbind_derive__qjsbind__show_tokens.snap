---
source: qjsbind-derive/src/qjsbind.rs
expression: "rustfmt_snippet::rustfmt(&patched.to_string()).unwrap()"
---
mod native_classes {
    use super::{KeyGenAlgorithm, Result};
    use js::IntoNativeObject as _;
    #[derive(js :: GcMark)]
    pub struct CryptoKey {}
    impl CryptoKey {
        pub fn new(inner: CryptoKey) -> Result<Self> {
            Ok(inner)
        }
    }
    mod qjsbind_generated {
        #![allow(non_snake_case)]
        use super::*;
        use js as crate_js;
        impl crate_js::Named for CryptoKey {
            const CLASS_NAME: &'static str = "CryptoKey";
        }
        impl crate_js::NativeClass for CryptoKey {
            fn constructor_object(ctx: &crate_js::Context) -> crate_js::Result<crate_js::Value> {
                ctx.get_qjsbind_object(std::any::type_name::<CryptoKey>(), || {
                    let constructor = ctx.new_function(
                        "CryptoKey",
                        qjsbind_CryptoKey_constructor,
                        0,
                        crate_js::c::JS_CFUNC_constructor,
                    );
                    let proto = ctx.new_object("CryptoKey");
                    constructor.set_property("prototype", &proto)?;
                    Ok(constructor)
                })
            }
        }
        #[crate_js::host_call(with_context)]
        fn qjsbind_CryptoKey_constructor(
            ctx: crate_js::Context,
            _this_value: crate_js::Value,
            inner: CryptoKey,
        ) -> crate_js::Result<crate_js::Native<CryptoKey>> {
            #[allow(unused_variables)]
            let ctx = ctx;
            use crate_js::IntoNativeObject;
            CryptoKey::new(inner).into_native_object(&ctx)
        }
    }
}
